File: app/src/main/java/com/michaelfotiadis/samplearchitecture/ui/posts/viewmodel/PostListCanceller.java
package com.michaelfotiadis.samplearchitecture.ui.posts.viewmodel;

import android.arch.lifecycle.Lifecycle;
import android.arch.lifecycle.LifecycleObserver;
import android.arch.lifecycle.OnLifecycleEvent;

import com.michaelfotiadis.samplearchitecture.domain.listener.Canceller;

public class PostListCanceller extends Canceller implements LifecycleObserver {

    private final Lifecycle lifecycle;

    public PostListCanceller(Lifecycle lifecycle) {
        this.lifecycle = lifecycle;
        this.lifecycle.addObserver(this);
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_STOP)
    void onStop() {
        cancel();
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_DESTROY)
    void onDestroy() {
        unRegister();
        lifecycle.removeObserver(this);
    }

}

File: app/src/main/java/com/michaelfotiadis/samplearchitecture/ui/posts/viewmodel/PostListCanceller.kt
package com.michaelfotiadis.samplearchitecture.ui.posts.viewmodel

import android.arch.lifecycle.Lifecycle
import android.arch.lifecycle.LifecycleObserver
import android.arch.lifecycle.OnLifecycleEvent

import com.michaelfotiadis.samplearchitecture.domain.listener.Canceller

class PostListCanceller(private val lifecycle: Lifecycle) : Canceller(), LifecycleObserver {

    init {
        this.lifecycle.addObserver(this)
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_STOP)
    internal fun onStop() {
        cancel()
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_DESTROY)
    internal fun onDestroy() {
        unRegister()
        lifecycle.removeObserver(this)
    }

}
